### PR TITLE
Remove unnecessary use of globals and global keyword

### PR DESCRIPTION
**Pull request by awwad**
*[Updated]*

The "global" keyword is only necessary when a module-level name is going
to be reassigned, at module level.

E.g. it's necessary here:
```
def fn():
  global x
  x = new_thing
```

But it is not generally necessary here:
```
def fn():
  x.some_method_that_changes_the_object()
```

Nor here:
```
def fn():
  some_method_that_changes_something(x)
```

Nor here:
```
def fn():
  x.element = new_thing
```

So I've removed its unnecessary use in such cases.

I also removed a global variable that was never used in the demo Secondary, and moved several initializations in `test_primary.py` out of separate functions and to module level (making global keywords unnecessary in further functions in `test_primary`).

### CODE REVIEW

**Comment by vladimir-v-diaz** [comment]
A few global variables were removed (presumably the ones that will potentially be reassigned), while others were not.  If possible, we should try to eliminate the use of globals entirely.

Do you think that storing the remaining global variables in a class will improve readability?  Keeping track of variables that are treated as globals, and that are later modified anywhere in the code, can be difficult for the reader to manage.  Many developers recommend that if your module does use a global variable, that it be set to a constant value.

Although a bit dated, [this](http://amir.rachum.com/blog/2013/07/09/python-common-newbie-mistakes-part-2/) article does a good job of showing the problems with globals.

**Comment by awwad** [comment]
I don't think classes would help here. I agree that globals are suboptimal. Here, they're in the demo code, which I worry a bit less about.

I can imagine having some dictionary in `demo/__init__.py` that stores the appropriate demo information for the demo modules, but that seems kinda clunky and confusing if you have multiple processes open (which you will in the demo). Maybe I'm wrong on that and I should just do the dictionary thing.

**Comment by awwad** [comment]
Hijacking this PR from myself for the broader purpose of the destruction of globals (global destruction?).

Removed the globals in test_primary by initializing the values in module-level code rather than inside functions.

That would slow importing test_primary down significantly, but realistically, things aren't going to import test_primary.

See commit 21639e6.

This would then be at module-level:
```python
# Load the private key for this Primary ECU.
primary_ecu_key = uptane.common.canonical_key_from_pub_and_pri(
    demo.import_public_key('primary'), demo.import_private_key('primary'))

# Generate a trusted initial time for the Primary.
clock = tuf.formats.unix_timestamp_to_datetime(int(time.time()))
clock = clock.isoformat() + 'Z'
tuf.formats.ISO8601_DATETIME_SCHEMA.check_match(clock)

# Load the public timeserver key.
key_timeserver_pub = demo.import_public_key('timeserver')
key_timeserver_pri = demo.import_private_key('timeserver')
```

This means that unittest method setUpModule will be empty:
```python
def setUpModule():
  """
  This is run once for the full module, before all tests.
  """
  destroy_temp_dir()
```

**Comment by vladimir-v-diaz** [comment]
I'm looking at the list of pull requests that require my review and this one shows up.  I noticed you replied and made some additional commits.  Do I have to review or respond?

**Comment by awwad** [comment]
I could use some thoughts on this, yes, before I go and repeat this sort of thing in other *test* modules. In particular, 21639e6 above (Remove all globals from test_primary), which does time-consuming initialization (key imports) at module level (during import, therefore, if it is ever imported)

**Comment by awwad** [comment]
Also, even after the commits, test_primary still uses one global: the Primary instance that the successive tests configure and mess with. Would you suggest that we find some way around that, too? (It seems weird to run `secondary::__init__` at module level when that's something we're explicitly testing in one of the unittest methods....)

File: demo/demo_secondary.py [hunk 1]
```diff
@@ -53,7 +53,6 @@
 _primary_host = demo.PRIMARY_SERVER_HOST
 _primary_port = demo.PRIMARY_SERVER_DEFAULT_PORT
 firmware_filename = 'secondary_firmware.txt'
-current_firmware_fileinfo = {}
```
Comment (lukpueh) [inline demo/demo_secondary.py:56]: I noticed that this variable still exists in [demo_primary](https://github.com/awwad/uptane/blob/21639e667be86d9d4741a9aa5560bf3aed4a4da7/demo/demo_primary.py#L80) but is never used.

**Review by lukpueh — COMMENTED** [review]
LGTM. Successfully ran passing unit tests, manually checked variables where `global` was removed.

 I would go and merge this as "removal of unnecessary global *keywords*" and create a separate PR to replace unnecessary global *variables*.

File: demo/demo_secondary.py (hunk 1 above)
Comment (awwad) [reply demo/demo_secondary.py:56]: Thanks. Fixed locally and will percolate upstream. (:

**Comment by awwad** [comment]
That would be a sensible separation, except that this PR actually does both, depending on exactly what you mean.

You could do three different things: remove an unnecessary global keyword, remove treatment of a variable as global, or remove a variable that happened to be global. This PR does the first two, and not the third. The first commit does the first in the demo. The second commit does the second in test_primary. If you refer to the third, then yes, that's not done here.

So anyway, I renamed the PR to indicate the first two.

**Comment by awwad** [comment]
Can you tell me what you think of my earlier comments on this PR, though?
[This one](https://github.com/uptane/uptane/pull/88#issuecomment-302444389) and [this one](https://github.com/uptane/uptane/pull/88#issuecomment-302447530).

(Basically: should I be doing time-consuming initialization at module level (on import) in a test module in order to avoid needing the `global` keyword?)

**Comment by lukpueh** [comment]
Re above questions... I don't have a strong argument to not do it your way. But, I personally use `unittest`'s `setUpClass` methods and define class variables instead of globals, because it seems more compliant and intuitive, especially if an error occurs during one of those setup calls `e.g. secondary::__init__`.

**Comment by awwad** [comment]
That is actually an excellent suggestion. >.< I definitely should have thought of that.

**Comment by awwad** [comment]
Closing. This is handled in the manner Lukas recommended (class variables) in another branch, to be PR'd soon.

**Pull request closed without merging**